src/luarocks/cmd.lua

```diff
@@ -306,6 +304,141 @@ do
+      :help_max_width(80)
+      :add_help("--help")
+      :add_help_command()
+      :add_complete_command({
+         help_max_width = 120,
+         summary = "Output a shell completion script.",
+         description = [[
+Output a shell completion script.
+
+Enabling completions for Bash:
+
+   Add the following line to your ~/.bashrc:
+      source <(]]..basename..[[ completion bash)
+   or save the completion script to the local completion directory:
+      ]]..basename..[[ completion bash > ~/.local/share/bash-completion/completions/]]..basename..[[
+
+
+Enabling completions for Zsh:
+
+   Save the completion script to a file in your $fpath.
+   You can add a new directory to your $fpath by adding e.g.
+      fpath=(~/.zfunc $fpath)
+   to your ~/.zshrc.
+   Then run:
+      ]]..basename..[[ completion zsh > ~/.zfunc/_]]..basename..[[
+
+
+Enabling completion for Fish:
+
+   Add the following line to your ~/.config/fish/config.fish:
+      ]]..basename..[[ completion fish | source
+   or save the completion script to the local completion directory:
+      ]]..basename..[[ completion fish > ~/.config/fish/completions/]]..basename..[[.fish
+]]})
+      :command_target("command")
+      :require_command(false)
+
+   parser:flag("--version", "Show version info and exit.")
+      :action(function()
+         util.printout(program.." "..cfg.program_version)
+         util.printout(description)
+         util.printout()
+         os.exit(cmd.errorcodes.OK)
+      end)
+   parser:flag("--dev", "Enable the sub-repositories in rocks servers for "..
+      "rockspecs of in-development versions.")
+   parser:option("--server", "Fetch rocks/rockspecs from this server "..
+      "(takes priority over config file).")
+   parser:option("--only-server", "Fetch rocks/rockspecs from this server only "..
+      "(overrides any entries in the config file).")
+      :argname("<server>")
+   parser:option("--only-sources", "Restrict downloads to paths matching the given URL.")
+      :argname("<url>")
+   parser:option("--namespace"):description("Specify the rocks server namespace to use.")
+   parser:option("--lua-dir", "Which Lua installation to use.")
+      :argname("<prefix>")
+   parser:option("--lua-version", "Which Lua version to use.")
+      :argname("<ver>")
+   parser:option("--tree", "Which tree to operate on.")
+   parser:flag("--local", "Use the tree in the user's home directory.\n"..
+      "To enable it, see '"..program.." help path'.")
+   parser:flag("--global", "Use the system tree when `local_by_default` is `true`.")
+   parser:flag("--verbose", "Display verbose output of commands executed.")
+   parser:option("--timeout", "Timeout on network operations, in seconds.\n"..
+      "0 means no timeout (wait forever). Default is "..
+      tostring(cfg.connection_timeout)..".")
+      :argname("<seconds>")
+      :convert(tonumber)
+   parser:option("-c"):description("Set the value of a config variable. "..
```

Comment:
This was my idea:

> ```
> 2019-06-05 13:20:03     pauloue the luarocks VAR=VALUE argument to set config variables is a bit difficult to do with argparse
> 2019-06-05 13:22:09     pauloue it's optional so if an argument is in argparse, it will consume the subcommand
> 2019-06-05 13:26:39     daurnimator     I don't understand your statement there?
> 2019-06-05 13:29:29     pauloue if the main command has an optional argument and a command, then the command will always be interpreted as an argument
> 2019-06-05 13:30:06     pauloue because argparse doesn't know when to stop parsing the arguments since there can be any number of them
> 2019-06-05 13:30:22     daurnimator     I thought that those config variables could be done with: parser:argument("configuration", "configuration overrides in NAME=VALUE form"):count("*")
> 2019-06-05 13:31:45     pauloue yes but then I think if you do luarocks install, the install will be interpreted as an config argument
> 2019-06-05 13:32:17     daurnimator     oh I was thinking that you'd add that configuration argument to the install subcommand; after the install target
> 2019-06-05 13:32:58     pauloue oh that would work
> 2019-06-05 13:34:27     pauloue does it apply to any other commands as well?
> 2019-06-05 13:34:40     daurnimator     I think build, install and make?
> 2019-06-05 13:34:47     daurnimator     need to check
> 2019-06-05 13:39:33     pauloue do you think it makes sense to drop the boolean argument for add_help_command in argparse? since it's not enabled by default
> 2019-06-05 13:40:37     pauloue I did that for add_complete to avoid tracking the option index since it could change if the help optional is removed
> 2019-06-05 13:42:40     daurnimator     I don't really have an opinion on it
> 2019-06-05 13:45:06     pauloue I'll do it then since it's simpler and there's no need for add_help_command(false)
> 2019-06-05 13:45:45     daurnimator     okay
> 2019-06-05 13:46:34     daurnimator     okay I was wrong about those variables... they're used in all sorts of minor ways all over the place.... urgh
> 2019-06-05 13:47:06     daurnimator     I guess every subcommand gets a trailing :argument("variable"):count("*")
> 2019-06-05 13:49:03     daurnimator     okay maybe not all of them..... this needs some closer auditing
> 2019-06-05 13:49:52     pauloue some subcommands already have an optional argument, like the rockspec for make
> 2019-06-05 13:51:20     pauloue a workaround could be to add a separate option for each config variable to the main command
> 2019-06-05 13:51:30     pauloue options don't have to start with a -
> 2019-06-05 13:51:37     daurnimator     yeah. I like git's -C for this sort of thing much better
> 2019-06-05 13:52:01     daurnimator     ==> with git you can do: git -C myvar.mysubvar=myvalue subcommand --etc
> 2019-06-05 13:52:15     daurnimator     uh I mean -c
> 2019-06-05 13:52:26     pauloue yes that be the simplest way
> 2019-06-05 13:52:26     daurnimator     -C is to change directories :P
> 2019-06-05 13:52:51     daurnimator     Sadly that's not something we can change easily though :(
> 2019-06-05 13:53:22     daurnimator     You might have to do a pre-argparse pass to extract the variables
> 2019-06-05 13:53:52     daurnimator     that's sort of what already happens in the luarocks cmdline parsing
> 2019-06-05 13:54:10     daurnimator     https://github.com/luarocks/luarocks/blob/daf7e7e5f5ede8c3a5a00e2edaf8246209712781/src/luarocks/cmd.lua#L319
> 2019-06-05 13:54:59     pauloue ok I'll do that
> 2019-06-05 13:55:51     daurnimator     Towards the end of your GSoC time, if we get time, we could *also* add -c like git
> 2019-06-05 13:56:04     daurnimator     and say that the VAR=NAME anywhere is deprecated
> 2019-06-05 13:56:11     daurnimator     but I'd want to give it a year or two before we remove it
> 2019-06-05 13:57:05     pauloue ok
> ```

Reply:
I have an issue with this feature, especially if it was made to mimic git:

In git, `-c foo=bar` is the temporary one-shot equivalent to setting `foo` to `bar` in `git config`.

But here, this is not equivalent to the same in `luarocks config`. It only applies to `luarocks config`'s `variables` subset. I don't like this inconsistency, I think it can be confusing/misleading.

I would rather remove the `"-c"` functionality for now from this PR, and consider bring it back later in the future as a `-c/--config` flag that applies to the entire configuration.

The current bare `VAR=VALUE` support was made to mimic Make, meaning `"$(VAR)"` value interpolations in the rockspec `build` section (and in retrospect should have allowed override via environment variables as well, as Make does).

Reply:
Okay. Sorry @pauloue. Could you remove the `-c` support for now?

Reply:
Dropped the commit that added it